azurelinuxagent/ga/telemetry_service.py

```diff
@@ -0,0 +1,146 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import threading
+import time
+import traceback
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.exception import ServiceStoppedError
+from azurelinuxagent.common.future import ustr, Queue, Full
+from azurelinuxagent.common.interfaces import ThreadHandlerInterface
+
+
+def get_telemetry_service_handler(protocol_util):
+    return TelemetryServiceHandler(protocol_util)
+
+
+class TelemetryServiceHandler(ThreadHandlerInterface):
```

Comment:
I didn't add it initially because this thread was just doing 1 single operation (which is not even a PeriodicOperation). On a deeper inspection, I realized I'm using certain things at a ThreadHandler level (like Queue, Event, etc) which might make the decoupling a bit more tedious. 

Given that the currently the thread is just doing one thing and purely from the perspective of not wanting to do extra work right now, I'm going to skip it but if any of you guys feel strongly about it let me know I'll change it :)

Reply:
When I first saw the use of SendTelemetryEventsHandler and CollectTelemetryEventsHandler from the main thread this threw me off. 

The main thread uses those handlers to manage (start/stop/query) the corresponding threads so I was surprised that one thread had a dependency on another. I had to dig more into the code to realize that CollectTelemetryEventsHandler doesn't really care about SendTelemetryEventsHandler's thread, but about the service that is running on that thread.

You can leave as is if you prefer so.

Reply:
haha yeah its all decoupled. Is there any comment or anything I can add to make it more clearer? To avoid someone else having the same confusion you had basically